libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -212,25 +212,94 @@ errorHandlers =
     ]
 {-# INLINE errorHandlers #-}
 
+-- | If the log level is less sensitive than 'Debug' just call the underlying app unchanged.
+-- Otherwise, pull a copy of the request body before running it, and if response status is @>=
+-- 400@, log the entire request, including the body.
+--
+-- The request sanitizer is called on the 'Request' and its body before it is being logged,
+-- giving you a chance to erase any confidential information.  Eg., if you're only interested
+-- in one end-point that has no sensitive information, use 'onlyLogEndpoint'.
+--
+-- WARNINGS:
+--
+--  * This may log confidential information if contained in the request.  Use the sanitizer to
+--    avoid that.
+--  * This does not catch any exceptions in the underlying app, so consider calling
+--    'catchErrors' before this.
+--  * Be careful with trying this in production: this puts a performance penalty on every
+--    request (unless level is less sensitive than 'Debug').
+heavyDebugLogging
+    :: ((Request, LByteString) -> Maybe (Request, LByteString))
+    -> Level -> Logger -> Middleware
+heavyDebugLogging sanitizeReq lvl lgr app = \req cont -> do
+    (bdy, req') <- if lvl <= Debug
```

Comment:
Maybe add a note here that this is for Debug or Trace; I had to go look up the `Ord` instance to figure out which direction was which :)

Reply:
I intentionally used the `Ord` instance because it is robust against changes in `Level`.  This is not really a practical concern, but still.  (I added a comment.)